Add an option to save without reproducibility

arbellea opened the pull request:
Hi,
I found that Tap().save() always saves with reproducibility info it is in a git repo.
However, this has two drawbacks:
1. Not everyone wants to save the reproducibility info.
2. In older versions of git (less than 2.0 I think) the code fails with an error.
I added two minor modifications which I hope you will accept.
1. Added a parameter to save: ```def save(self, path: str, with_reproducibility: bool = True)```
2. Added  a try/except around the git part to avoid code crash due to old git version

arbellea committed:
Add option not to save reproducability info

Also added a try/except around the git part since it seems to fail on older git version.

arbellea committed:
Merge pull request #1 from arbellea/arbellea-patch-1

Add option not to save reproducibility info

martinjm97 commented:
I'm a big fan of: `def save(self, path: str, with_reproducibility: bool = True)`. I thought about this when we were discussing configs. I agree both with your choice to add this flag and the default behavior of keeping the reproducibility information. 

With regard to the `try`/`except`, I'd prefer that the `expect` was scoped to a specific exception. Did you encounter a particular error with an old version of git? The helpers called in `get_reproducibility_info` already handle some exceptions.

Thank you for contributing to Tap! With the change to the blank except, it should be a nice improvement.

arbellea committed:
Added specific except to reproducability

The error occurs on "git remote get-url origin" when git version is less than 2.0
Added 
```from subprocess import CalledProcessError```
and ```except CalledProcessError```

arbellea commented:
Hi, I added the except to catch the error raised when git version <2.0

martinjm97 commented:
Hi @arbellea,

Sorry for so much back and forth. We're having a little bit of trouble testing with git version <2.0. We're curious if it's possible to push the error checking (the `try` `except CalledProcessError`) inside `has_git` to keep the code in `save` as clean as possible. So we're wondering, what function the error is coming from? Thank you very much for looking into this! Definitely an issue we wouldn't have discovered.

arbellea committed:
Move try/except into utils. 

In except, use command that works with old git versions.

arbellea committed:
Remove try/except from reproducibility

Moved into utils.py

arbellea commented:
Hi,
I found the error and the alternate command that works in the older git version.
Moved the try/except into utils.py

martinjm97 reviewed (approved):
Overall, this PR is excellent! Much appreciated. Comments corresponding to small changes that I'll make:
1) `with_reproducibility` needs a docstring.
2) A minor and subjective code refactor. 
`args = self._log_all() if with_reproducibility else self.as_dict()`
`json.dump(args, f, indent=4, sort_keys=True, cls=PythonObjectEncoder)`

Thanks!

arbellea commented:
Great, Thanks
When the pip version will update, I'll also update the code in my projects!

martinjm97 commented:
Thanks again! The version update will be up in the next week or so (unless you need sooner for some reason). We're hoping to finally resolve [this tricky issue](https://github.com/swansonk14/typed-argument-parser/issues/12) before the next release!

arbellea commented:
No rush.... 
It does look tricky :)

martinjm97 commented:
@arbellea Release is out [v1.5.1](https://github.com/swansonk14/typed-argument-parser/releases/tag/v_1.5.1)!

arbellea commented:
Thanks! 😄
